ipn/ipnlocal/local.go

```diff
@@ -624,58 +624,96 @@ func (b *LocalBackend) SetDirectFileRoot(dir string) {
 //
 // It returns (false, nil) if not running in declarative mode, (true, nil) on
 // success, or (false, error) on failure.
+//
+// If it is called whilst the LocalBackend is in 'Running' state, only a subset
+// of changes currently deemed safe to chnage for a running system are applied.
+// The rest of the changes will get applied next time LocalBackend starts and
+// config is reloaded. Currently (08/2024) changes safe to apply while running
+// are changes to '.AcceptRoutes', '.AdvertiseRoutes' and '.StaticEndpoints'
+// fields.
 func (b *LocalBackend) ReloadConfig() (ok bool, err error) {
-	b.mu.Lock()
-	defer b.mu.Unlock()
+	unlock := b.lockAndGetUnlock()
+	defer unlock()
 	if b.conf == nil {
 		return false, nil
 	}
 	conf, err := conffile.Load(b.conf.Path)
 	if err != nil {
 		return false, err
 	}
-	if err := b.setConfigLocked(conf); err != nil {
+	if err := b.setConfigLocked(conf, unlock); err != nil {
 		return false, fmt.Errorf("error setting config: %w", err)
 	}
 
 	return true, nil
 }
 
-func (b *LocalBackend) setConfigLocked(conf *conffile.Config) error {
-
-	// TODO(irbekrm): notify the relevant components to consume any prefs
-	// updates. Currently only initial configfile settings are applied
-	// immediately.
-	p := b.pm.CurrentPrefs().AsStruct()
-	mp, err := conf.Parsed.ToPrefs()
-	if err != nil {
-		return fmt.Errorf("error parsing config to prefs: %w", err)
-	}
-	p.ApplyEdits(&mp)
-	if err := b.pm.SetPrefs(p.View(), ipn.NetworkProfile{}); err != nil {
-		return err
-	}
-
+// setConfigLocked applies tailscaled config from conffile.Config.
+// unlock must be non-nil if called for a 'Running' LocalBackend.
+func (b *LocalBackend) setConfigLocked(conf *conffile.Config, unlock unlockOnce) error {
 	defer func() {
 		b.conf = conf
 	}()
 
-	if conf.Parsed.StaticEndpoints == nil && (b.conf == nil || b.conf.Parsed.StaticEndpoints == nil) {
-		return nil
-	}
-
 	// Ensure that magicsock conn has the up to date static wireguard
 	// endpoints. Setting the endpoints here triggers an asynchronous update
 	// of the node's advertised endpoints.
-	if b.conf == nil && len(conf.Parsed.StaticEndpoints) != 0 || !reflect.DeepEqual(conf.Parsed.StaticEndpoints, b.conf.Parsed.StaticEndpoints) {
+	if b.needStaticEndpointsUpdate(conf) {
 		ms, ok := b.sys.MagicSock.GetOK()
 		if !ok {
 			b.logf("[unexpected] ReloadConfig: MagicSock not set")
 		} else {
 			ms.SetStaticEndpoints(views.SliceOf(conf.Parsed.StaticEndpoints))
 		}
 	}
-	return nil
+
+	oldPrefs := b.pm.CurrentPrefs().AsStruct()
+	newPrefs, err := conf.Parsed.ToPrefs()
+	if err != nil {
+		return fmt.Errorf("error parsing config to prefs: %w", err)
+	}
+
+	// If we are not running it should be safe to set all prefs. We can
+	// assume that these changes will get propagated before state changes to
+	// 'Running'.
+	if b.state != ipn.Running {
```

Comment:
I'm assuming that the configfile will really only be used in container-like environments where the state change sequence is a bit simpler (i.e 'starting by running `tailscaled --config <configfile-path>`, auth key passed in configfile' -> 'Running' -> 'the whole thing goes down' -> 'starting by running `tailscaled ..`. I haven't thought about what would happen if someone calls reload when the state is 'Stopped' if that's possible, but not sure if we need to worry about that.

Reply:
I guess I don't get why we need to do different things in Running state vs not running state. Why can't we do the same thing on Reload regardless?

That "selectively apply a few known safe prefs" worries me. What's an unsafe pref?

Reply:
> I guess I don't get why we need to do different things in Running state vs not running state. Why can't we do the same thing on Reload regardless?

I guess because it seemed to me that there isn't a single existing flow that we could trigger on Reload that would tell each component to read the new state from prefs as it happens on start?
Right now, if Reload is called when the state is 'Running' it sets prefs, but doesn't actually update the system (i.e doesn't update filter to add any new routes that need to be advertized etc) which seems double suboptimal because an arbitrary change down the line that _would_ update filter would also pick up the change from Reload config.
And because there are a lot of things that feels like would require thought to be able to reset safely (i.e changing netfilter mode or control server URL), it felt like we can't get there with a single PR either?